### PR TITLE
Upgrade Catch2 to version 3+

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Preparation to replace `RunTests` with one test exec for each library.

The two fixup commits are intended to help reviewers since they do a simple find-replace. Note also that the container doesn't have Catch2 v3 installed yet.

### Upgrade instructions

- Building tests now requires Catch2 version 3+. Install with your package manager (`brew install catch2` on macOS) or with a standard CMake build and install from source (see the [Catch2 docs](https://github.com/catchorg/Catch2/blob/devel/docs/cmake-integration.md#installing-catch2-from-git-repository)). Build with `BUILD_SHARED_LIBS=ON` to build shared libs, or with `CMAKE_POSITION_INDEPENDENT_CODE=ON` if you're building static libs.
- Rename `Catch::Contains` to `Catch::Matchers::ContainsSubstring`.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: docs/DevGuide/BuildSystem.md [hunk 1]
```diff
@@ -142,11 +142,11 @@ dev_guide_creating_executables "how to create executables".
 
 To add an external dependency, first add a `SetupDEPENDENCY.cmake`
 file to the `cmake` directory. You should model this after one of the
-existing ones for `Catch` or `Brigand` if you're adding a header-only
+existing one for `Brigand` if you're adding a header-only
```
Comment (wthrowe) [inline docs/DevGuide/BuildSystem.md:145]: "after the existing one"

File: tests/Unit/Parallel/Test_AlgorithmCore.cpp [hunk 2]
```diff
@@ -2,7 +2,6 @@
 // See LICENSE.txt for details.
 
 // Need CATCH_CONFIG_RUNNER to avoid linking errors with Catch2
-#define CATCH_CONFIG_RUNNER
```
Comment (wthrowe) [inline tests/Unit/Parallel/Test_AlgorithmCore.cpp:5]: Remove dangling comment here and in several other places.

**Review by wthrowe — CHANGES_REQUESTED** [review]
Haven't tested it, but looks reasonable.

**Review by kidder — DISMISSED** [review]
When building with on my desktop with clang16 and `BUILD_SHARED_LIBS=ON` I am getting multiple undefined `Catch::Approx` symbols in tests that have their own executable (e.g. Test_AlgorithmParallel.)  RunTests builds fine.

**Comment by nilsvu** [comment]
Thanks for checking @kidder! I'll take a look at those.

**Comment by nilsvu** [comment]
This is ARM macOS?

**Comment by kidder** [comment]
> This is ARM macOS?

No this is a linux desktop with Fedora38, catch version 3.3.2

**Comment by nilsvu** [comment]
Was an easy fix. I also updated the Dockerfile but haven't built or pushed the container yet because it takes so long to build on my ARM mac (it has to emulate x86).

**Comment by kidder** [comment]
> Was an easy fix. I also updated the Dockerfile but haven't built or pushed the container yet because it takes so long to build on my ARM mac (it has to emulate x86).

you need a similar fix in `tests/Unit/IO/Importers/CMakeLists.txt`

**Review by kidder — COMMENTED** [review]
LGTM

**Comment by nilsvu** [comment]
Container is now pushed and fixups squashed. CI failures are timeouts.

**Review by wthrowe — APPROVED** [review]

**Review by nilsdeppe — COMMENTED** [review]
Seems to mostly work. I get segfaults in:
```
 627 - Unit.Evolution.Systems.GeneralizedHarmonic.BCBjorhus.Cls (Failed)
 636 - Unit.Evolution.Systems.GeneralizedHarmonic.Characteristics (Failed)
 644 - Unit.GhValenciaDivClean.BoundaryConditions.ConstraintPreservingFreeOutflow (Failed)
 647 - Unit.GhValenciaDivClean.BoundaryConditions.ProductOfConditions (Failed)
```

**Comment by nilsvu** [comment]
Could you post backtraces or do you understand what's segfaulting there?

**Comment by nilsdeppe** [comment]
Heh, ignore, something went weird on my machine

**Comment by nilsvu** [comment]
Ok, ready to get this in? Then I can open the next PR that splits up RunTests.

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**